Implemented game loop

carterjmoore opened the pull request:
Set up the game loop. With every loop, we get a random instruction from a list of instructions and then add the corresponding UI elements. If the proper input is received, we clear the UI elements, increase score, wait one second, and then loop again. Also made it so the time available to perform the actions scales with score, but we might want to adjust this scale in the future.

carterjmoore committed:
Implemented game loop

Set up the game loop. With every loop, we get a random instruction from a list of instructions and then add the corresponding UI elements. If the proper input is received, we clear the UI elements, wait one second, and then loop again. Also made it so the time available to perform the actions scales with score, but we might want to adjust this scale in the future.

fredy20501 commented on app/src/main/java/ca/unb/mobiledev/reflexrevolution/GameActivity.java:
```diff
@@ -0,0 +1,162 @@
+        resetTimer();
+    }
+
+    //Fill the instructions list with instructions
+    private void createInstructions(){
+        instructions = new ArrayList<>();
+
+        //Add instructions here
+        instructions.add("Button"); //Temp instruction for testing
+    }
```
This is good for now, but later on the list of instructions will be given by the main menu (depending on what game mode the player selected) so it will actually take the list from the data bundle passed to this activity.

We can do this later or, if you want, you can implement that functionality in this PR (define the ArrayList in MainActivity and pass it as extras with the Intent to GameActivity).

carterjmoore replied:
Do you think it would also be possible to just pass in a string for "GameMode" from the MainActivity, and then build the list of instructions in GameActivity based on that game mode? Or would it be better to just pass the ArrayList right from MainActivity?

fredy20501 replied:
Yeah, that could definitely work too. Just passing a string would make it a bit simpler, and whether we keep the lists in the MainActivity or the GameActivity doesn't really make a difference.

fredy20501 commented on app/src/main/java/ca/unb/mobiledev/reflexrevolution/GameActivity.java:
```diff
@@ -0,0 +1,162 @@
+        }.start();
+    }
+
+    //Create a new timer that counts down from 5, and closes the activity once it hits 0
+    private void newTimer(){
+        timeCount = scaleTimerFromScore();
+        timer = new CountDownTimer(timeCount, 100) {
```
It might be good to make sure the timer isn't already running before creating a new one. Something similar to what is done in `resetTimer()` should work:
```java
if(timer != null) {
    timer.cancel();
}
```

carterjmoore replied:
Alright, I'll add that just to be safe

fredy20501 commented on app/src/main/java/ca/unb/mobiledev/reflexrevolution/GameActivity.java:
```diff
@@ -0,0 +1,162 @@
+        newTimer();
+    }
+
+    //Returns a random instruction from instructions
+    private String getRandomInstruction() {
+        Random rand = new Random();
```
We can move the `new Random()` initialization into the `onCreate()` method so we're not re-initializing a new random number generator each time (keep it as a private class variable).

carterjmoore replied:
Good idea! I'll throw that in :)

fredy20501 commented on app/src/main/java/ca/unb/mobiledev/reflexrevolution/GameActivity.java:
```diff
@@ -0,0 +1,162 @@
+    private LinearLayout layout; //Layout we should add new UI elements to
+
+    private CountDownTimer timer;
+    private String currentInstruction = null;
+    private ArrayList<String> instructions;
```
I think defining a custom enum would make the code clearer than using strings. We can simply define the public enum in its own file. Then we can use this new `Instruction` type instead of string, and use a nice `switch()` statement instead of a bunch of if-else inside the `displayInstruction()` method.
```java
package ca.unb.mobiledev.reflexrevolution;

public enum Instruction {
    NONE,
    TAP,
    SWIPE,
    SHAKE
    // Will add more later
}
```
```java
switch(currentInstruction) {
    case TAP:
        // Do stuff
        break;
    case SWIPE:
        // Do stuff
        break;
    case SHAKE:
        // Do stuff
        break;
}
```

carterjmoore replied:
Yeah I like the idea of an enum better too, good plan!

fredy20501 commented on app/src/main/java/ca/unb/mobiledev/reflexrevolution/GameActivity.java:
```diff
@@ -0,0 +1,162 @@
+    private int timeCount;
+    private int score;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
```
We need to add the other lifecycle methods as well and make sure that both timers will start/stop/pause/restart at the appropriate time to prevent issues/unknown behaviour. 

Example:
Currently, if you start a game then use the device back button to go back to the main menu, the timer is still going in the background and you'll be redirected to the game over screen after a few seconds. Stoping the timers `onDestroy` should fix this particular case, I think.

carterjmoore replied:
I'm not sure what you mean about the onDestroy thing... do you think it would work if I just paused the timer onPause() and then resumed it onResume()?

fredy20501 replied:
Yeah so I don't really know the specific of how the countdown timer works so just pausing/resuming in the onPause() and onResume() might be enough. The onDestroy() method is called when the app is closed, so at that point the timer should definitely be stopped (I don't know if just pausing it might make it stay in memory and cause memory leaks, but I think that is unlikely)

carterjmoore committed:
Implemented changes Fred requested

Added enums for Instructions, GameMode, and Difficulty. Instruction list is now constructed based on GameMode. Timer now pauses when app is paused and resumed when it resumes. Also cleaned up a few other things suggested by Fred.

carterjmoore committed:
Merge branch 'main' into game-loop-timer

carterjmoore committed:
Fixed a bug where "Play Again" caused crashing

carterjmoore committed:
Merge branch 'game-loop-timer' of https://github.com/fredy20501/reflex-revolution into game-loop-timer

carterjmoore committed:
Swapped MainActivity to be the main activity

fredy20501 reviewed (approved):
Nicely done! 🥇

hsavoie commented:
Looks good! Nice work :)
